### ISSUE TITLE
AV in mscorlib_ni!System.Diagnostics.Tracing.ActivityTracker.OnStart

### ISSUE THREAD

**Issue by KKhurin** [dotnet/coreclr#2773]
This is the latest dnx-coreclr-win-x64.1.0.0-rc2-16357
We're getting AV on every event (which seem to be swallowed and ignored). 
Here is a typical call stack:
~~~
 # Child-SP          RetAddr           Call Site
00 00000091`2c64d9e0 00007ffd`12249f2f mscorlib_ni!System.Diagnostics.Tracing.ActivityTracker.OnStart(System.String, System.String, Int32, System.Guid ByRef, System.Guid ByRef, System.Diagnostics.Tracing.EventActivityOptions)+0xba
01 00000091`2c64da80 00007ffd`12249454 mscorlib_ni!System.Diagnostics.Tracing.EventSource.WriteEventWithRelatedActivityIdCore(Int32, System.Guid*, Int32, EventData*)+0x1af
02 00000091`2c64dbf0 00007ffc`b315143f mscorlib_ni!System.Diagnostics.Tracing.EventSource.WriteEvent(Int32, System.String)+0x94
03 00000091`2c64dc70 00007ffc`b3150c6d System_Private_ServiceModel!System.Runtime.WcfEventSource.ServiceChannelCallStart(System.Runtime.Diagnostics.EventTraceActivity)+0x3f
04 00000091`2c64dcb0 00007ffc`b3150691 System_Private_ServiceModel!System.ServiceModel.Channels.ServiceChannel.Call(System.String, Boolean, System.ServiceModel.Dispatcher.ProxyOperationRuntime, System.Object[], System.Object[], System.TimeSpan)+0xed
05 00000091`2c64ddf0 00007ffc`b314fb88 System_Private_ServiceModel!System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(System.ServiceModel.Channels.MethodCall, System.ServiceModel.Dispatcher.ProxyOperationRuntime)+0x61
06 00000091`2c64de60 00007ffd`12b6aab3 System_Private_ServiceModel!System.ServiceModel.Channels.ServiceChannelProxy.Invoke(System.Reflection.MethodInfo, System.Object[])+0xa8
~~~

This looks like a jitter bug to me 'cause we're missing a null check in the assembler code:
~~~
call    mscorlib_ni!System.Diagnostics.Tracing.ActivityTracker.Enable() // Enable();
// we're missing this code here: 
// if (m_current == null)
//  return;
// and we directly proceed with m_current.Value;
mov     ecx,dword ptr [esi+4]
cmp     dword ptr [ecx],ecx
call    mscorlib_ni!System.Threading.AsyncLocal`1[].get_Value() 
~~~

**Comment by sivarv** [comment]
@KKhurin - can you provide more context on the null check that JIT is missing to generate?

Note that in the assembly code that you have pasted, 'cmp' instruction is the null-check on thisptr on which method call is being made. 

mov     ecx,dword ptr [esi+4]
cmp     dword ptr [ecx],ecx
call    mscorlib_ni!System.Threading.AsyncLocal`1[].get_Value()

If the method call needs to take place on an obj-ref, only if it not null, then one should explicitly write code in C#, JIt will not automatically generate that for you.

Here is the relevant IL of the method:

IL_0046:  ldarg.0
IL_0047:  ldfld      class System.Threading.AsyncLocal`1<class System.Diagnostics.Tracing.ActivityTracker/ActivityInfo> System.Diagnostics.Tracing.ActivityTracker::m_current
IL_004c:  callvirt   instance !0 class System.Threading.AsyncLocal`1<class System.Diagnostics.Tracing.ActivityTracker/ActivityInfo>::get_Value()

To me this seems like a bug either in OnStart() method or the code that is using ActivityTracker has not initialized it properly to ensure that its m_current field is not null.

**Comment by mikedn** [comment]
I think he's talking about the second `m_current == null` check in this code:

```C#
            if (m_current == null)        // We are not enabled
            {
                // We  used to rely on the TPL provider turning us on, but that has the disadvantage that you don't get Start-Stop tracking
                // until you use Tasks for the first time (which you may never do).   Thus we change it to pull rather tan push for whether
                // we are enabled.   
                if (m_checkedForEnable)
                    return;
                m_checkedForEnable = true;
                if (TplEtwProvider.Log.IsEnabled(EventLevel.Informational, TplEtwProvider.Keywords.TasksFlowActivityIds))
                    Enable(); // the call to Enable can be seen in the disassembly
                if (m_current == null) // but this one seems to be missing
                    return;
            }

            Contract.Assert((options & EventActivityOptions.Disable) == 0);

            var currentActivity = m_current.Value;
```
But the disassembly dump is from a x86 build, it can't be dnx-coreclr-win-x64.1.0.0-rc2-16357 since that's a x64 build.

**Comment by KKhurin** [comment]
yep, the second check for null is missing
~~~
                if (m_current == null)
                    return;
~~~

I checked both x86 and x64 and to my surprise the problem is identical

**Comment by davmason** [comment]
This is not actually a jit issue. There was a bug where the second null check was missing, it was fixed in EventSource about 2 weeks ago but dnx-coreclr-win-x64.1.0.0-rc2-16357 is 3 weeks old so it's on the old sources with the bug. It was fixed with efa1dda2c1579417dcaed92633b8556fd68270ee

**Comment by hoyMS** [comment]
I didn't see the bad code gen with today's build by running "crossgen mscorlib.dll".

IN001a: 0000B6 mov      rcx, rsi
IN001b: 0000B9 call     System.Diagnostics.Tracing.ActivityTracker:Enable():this

G_M25575_IG07:        ; offs=0000BEH, size=0007H, gcrefRegs=000000C8 {rbx rsi rdi}, byrefRegs=0000C000 {r14 r15}, byref, isz

IN001c: 0000BE cmp      gword ptr [rsi+8], 0
IN001d: 0000C3 jne      SHORT G_M25575_IG09

It would be very helpful if you can provide a repro.
